File: J9Ch/src/S_3_Data_Structure_II/Optional_13/_370_Convert_Expression_to_Reverse_Polish_Notation_Hard.py

class Stack:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return len(self.items)==0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        if not self.isEmpty():
            return self.items[-1]

    def size(self):
        return len(self.items)

class Solution:
    # @param expression: A string list
    # @return: The Reverse Polish notation of this expression
    def getLevel(self, s):
        if s == "+" or s == "-":
            return 1
        if s == "*" or s == "/":
            return 2

        return 0

    def convertToRPN(self, expression):
        # write your code here
        RPN = []
        cal = Stack()
        for s in expression:
            if s == "(":
                cal.push(s)
            elif s == ")":
                while not cal.isEmpty() and cal.peek() != "(":
                    RPN.append(cal.peek())
                    cal.pop()
                cal.pop()
            elif s.isdigit():
                RPN.append(s)
            else:
                if not cal.isEmpty():
                    if cal.peek() != "(":
                        while self.getLevel(cal.peek()) >= self.getLevel(s):
                            RPN.append(cal.peek())
                            cal.pop()
                            if cal.isEmpty():
                                break
                cal.push(s)

        while not cal.isEmpty():
            RPN.append(cal.peek())
            cal.pop()

        return RPN

File: J9Ch/src/S_3_Data_Structure_II/Optional_13/python_cpp/_370_Convert_Expression_to_Reverse_Polish_Notation_Hard.cpp


class Solution {
public:
    /**
     * @param expression: A string array
     * @return: The Reverse Polish notation of this expression
     */
    int getLevel(string opt) {
        if (opt == "(")
            return 0;
        if (opt == "+" || opt == "-")
            return 1;
        if (opt == "*" || opt == "/")
            return 2;

        return 3;
    }

    bool isOperator(string c) {
        return (c == "+" || c == "-" || c == "*" || c == "/");
    }

    vector<string> convertToRPN(vector<string> &expression) {
        // write your code here
        stack<string> st;
        vector<string> RPN;
        int len = expression.size();
        for (int i = 0; i < len; ++i) {
            string c = expression[i];
            if (c == "(")
                st.push(c);
            else if (c == ")") {
                while (st.top() != "(") {
                    RPN.push_back(st.top());
                    st.pop();
                }
                st.pop();
            } else {
                if (!isOperator(c))
                    st.push(c);
                else {
                    while (!st.empty() && getLevel(st.top()) >= getLevel(c)) {
                            RPN.push_back(st.top());
                            st.pop();
                    }
                    st.push(c);
                }
            }
        }

        while (! st.empty()) {
            RPN.push_back(st.top());
            st.pop();
        }

        return RPN;
    }
};